flaml/searcher/online_searcher.py

```diff
@@ -0,0 +1,371 @@
+        space = sorted(list(itertools.combinations(
+                       champion_all_combinations, interaction_num_to_add)))
+        self._random_state.shuffle(space)
+        candidate_configs = [set(champion_config) | set(item) for item in space]
+        assert len(candidate_configs) <= len(champion_config) * (len(champion_config) - 1) / 2.0
+        final_candidate_configs = []
+        for c in candidate_configs:
+            new_c = set([e for e in c if len(e) > 1])
+            final_candidate_configs.append(new_c)
+        return final_candidate_configs
+
+    def _generate_num_candidates(self, config, num):
+        """Use local search to generate new candidate
+        """
+        self._random = np.random.RandomState(self.NUM_RANDOM_SEED)
+        logger.info('num %s', num)
+        half_num = int(num / 2)
+        new_candidate_list = []
+        hp_vec = list(config.keys())
+        hp_config_vec = [config[k] for k in hp_vec]
+        dim = len(hp_config_vec)
+        candidate_config_vecs = []
+        for i in range(half_num):
+            move_direction = self._rand_vector_unit_sphere(dim)
+            c1 = [hp_config_vec[i] * (2**move_direction[i]) for i in range(dim)]
+            c2 = [hp_config_vec[i] * (2**(-move_direction[i])) for i in range(dim)]
+            projected_c1 = self._project(hp_vec, c1)
+            projected_c2 = self._project(hp_vec, c2)
+            logger.info('on_trial_complete(trial.trial_id) %s %s %s', projected_c1,
+                        type(projected_c1), hp_config_vec)
+            if not (projected_c1 == hp_config_vec).all():
+                candidate_config_vecs.append(projected_c1)
+            if not (projected_c2 == hp_config_vec).all():
+                candidate_config_vecs.append(projected_c2)
+
+        for c in candidate_config_vecs:
+            new_config = {}
+            for i in range(dim):
+                new_config[hp_vec[i]] = c[i]
+            new_candidate_list.append(new_config)
+        logger.info('new_candidate_list %s %s', config, new_candidate_list)
+        return new_candidate_list
+
+    def _rand_vector_unit_sphere(self, dim) -> np.ndarray:
+        vec = self._random.normal(0, 1, dim)
+        mag = np.linalg.norm(vec)
+        return vec / mag
+
+    def _project(self, hp_key_vec, hp_value_vec):
+        projected_vec = []
+        for i in range(len(hp_key_vec)):
+            config_domain = self._space[hp_key_vec[i]]
+            logger.info('Float %s', config_domain)
+            value = hp_value_vec[i]
+            upper = config_domain.upper
+            lower = config_domain.lower
+            logger.info('config_domain %s', lower)
+            print(config_domain, lower, type(lower))
+            projected = min(max(value, lower), upper)
+            projected_vec.append(projected)
+        return np.array(projected_vec)
+
+    @staticmethod
+    def _generate_all_comb(champion_config, order=2):
+        def convert_nested_tuple_to_tuple(test_tuple):
```

Comment:
Is this function needed? Can't we flatten the tuple when generating the combination?